drivers/avr/apa102.c

```diff
@@ -21,76 +21,106 @@
     // and adapted. The code is MIT licensed. I think thats compatible?
-
-    // We need to send some more bytes to ensure that all the LEDs in the
-    // chain see their new color and start displaying it.
     //
     // The data stream seen by the last LED in the chain will be delayed by
     // (count - 1) clock edges, because each LED before it inverts the clock
     // line and delays the data by one clock edge.  Therefore, to make sure
     // the last LED actually receives the data we wrote, the number of extra
     // edges we send at the end of the frame must be at least (count - 1).
-    // For the APA102C, that is sufficient.
     //
-    // The SK9822 only updates after it sees 32 zero bits followed by one more
-    // rising edge.  To avoid having the update time depend on the color of
-    // the last LED, we send a dummy 0xFF byte.  (Unfortunately, this means
-    // that partial updates of the beginning of an LED strip are not possible;
-    // the LED after the last one you are trying to update will be black.)
-    // After that, to ensure that the last LED in the chain sees 32 zero bits
-    // and a rising edge, we need to send at least 65 + (count - 1) edges.  It
-    // is sufficent and simpler to just send (5 + count/16) bytes of zeros.
+    // Assuming we only want to send these edges in groups of size K, the
+    // C/C++ expression for the minimum number of groups to send is:
     //
-    // We are ignoring the specification for the end frame in the APA102/SK9822
-    // datasheets because it does not actually ensure that all the LEDs will
-    // start displaying their new colors right away.
-
-    apa102_send_byte(0xFF);
-    for (uint16_t i = 0; i < 5 + leds / 16; i++) {
+    //   ((count - 1) + (K - 1)) / K
+    //
+    // The C/C++ expression above is just (count - 1) divided by K,
+    // rounded up to the nearest whole number if there is a remainder.
+    //
+    // We set K to 16 and use the formula above as the number of frame-end
+    // bytes to transfer.  Each byte has 16 clock edges.
+    //
+    // We are ignoring the specification for the end frame in the APA102
+    // datasheet, which says to send 0xFF four times, because it does not work
+    // when you have 66 LEDs or more, and also it results in unwanted white
+    // pixels if you try to update fewer LEDs than are on your LED strip.
+    uint16_t iterations = (num_leds + 14) / 16;
+    for (uint16_t i = 0; i < iterations; i++) {
         apa102_send_byte(0);
     }
+
+    apa102_init();
 }
 
-void apa102_send_byte(uint8_t byte) {
-    uint8_t i;
-    for (i = 0; i < 8; i++) {
-        writePin(RGB_DI_PIN, !!(byte & (1 << (7 - i))));
-        writePinHigh(RGB_CLK_PIN);
-    }
+#define APA102_SEND_BIT(byte, bit)               \
+    do {                                         \
+        writePin(RGB_DI_PIN, (byte >> bit) & 1); \
+        writePinHigh(RGB_CLK_PIN);               \
+        writePinLow(RGB_CLK_PIN);                \
```

Comment:
This will usually be too fast on ARM devices, and will need a delay. Bitbanging pulses shows no activity using an oscilloscope, seemingly because of the peripheral clocks running slower than the main system clock.

Suggest adding a slight delay after each of the pin changes.

Reply:
Thanks for the input! Programming on a microcontroller is relatively new to me. How long of a delay do you suggest?

Reply:
@tzarc isn't this on the avr specific folder? 
The ARM one would be in the chibiOS folder, right?

Reply:
Since this is a bit-banging driver, it would be interesting to move it and make it a generic driver for both platforms, as all hardware access is already abstracted through QMKs GPIO controls

I don't have any (working) pro micros or other AVR dev boards right now, but I've tested this PR (with some makefile additions) and it's working on ARM https://www.youtube.com/watch?v=gNS1yyRnPxo
That's without a delay between the pin writes for the clock btw. I think I'd still add a few microseconds of delay to be sure tho

Reply:
@Duckle29 Thank you for testing this. I finally got my hand on an ARM microcontroller to test this out, running at 72 MHz it does look like a delay is necessary. From testing, a delay of 40 ns seems to be the minimum the LEDs can handle. I just pushed a906bed114c4919ec79e8ebecfe9cbc1412a9f75 with modifications.